src/serde/de.rs

```diff
@@ -0,0 +1,931 @@
+                .ok_or_else(|| IonError::decoding_error("expected a char, found an empty string"))
+        })?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_str(self.reader.read_str()?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_str(self.reader.read_str()?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_bytes<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_bytes(self.reader.read_blob()?.as_slice());
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_byte_buf<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_byte_buf(self.reader.read_blob()?.as_slice().to_vec());
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        if self.reader.is_null() {
+            // since this is is a null value which is equivalent to `None` in Rust data type,
+            // we can skip reading the next value and call visit_none
+            self.reader.next()?;
+            visitor.visit_none()
+        } else {
+            // For non null value we transfer the call to visit_some to perform the deserialization
+            visitor.visit_some(self)
+        }
+    }
+
+    fn deserialize_unit<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        if self.reader.is_null() {
+            self.reader.next()?;
+            visitor.visit_unit()
+        } else {
+            IonResult::decoding_error("expected a null value")
+        }
+    }
+
+    fn deserialize_unit_struct<V>(
+        self,
+        _name: &'static str,
+        visitor: V,
+    ) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        self.deserialize_unit(visitor)
+    }
+
+    fn deserialize_newtype_struct<V>(
+        self,
+        name: &'static str,
+        visitor: V,
+    ) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        if name == TUNNELED_TIMESTAMP_TYPE_NAME {
+            let timestamp = self.reader.read_timestamp()?;
+            assert_eq!(
+                std::mem::size_of::<V::Value>(),
+                std::mem::size_of::<Timestamp>()
+            );
+            // compiler doesn't understand that the generic Timestamp here is actually V::Value here
+            let visitor_value =
+                unsafe { std::mem::transmute_copy::<Timestamp, V::Value>(&timestamp) };
```

Comment:
[By convention](https://std-dev-guide.rust-lang.org/policy/safety-comments.html), any use of `unsafe` is accompanied by a `// SAFETY:` comment that explains why you're using `unsafe` (which you've got), why what you're doing is actually safe even though the compiler disagrees, and what other invariants/assumptions others need to ensure for it to keep working as intended.